File: src/resolver/manifest.ts
import { promises as fs } from "node:fs";
import * as path from "node:path";

import { InvalidManifestError } from "../errors/index.js";

import type { ProfileManifest, ResolutionWarning } from "./types.js";

const KNOWN_FIELDS = new Set([
  "name",
  "description",
  "extends",
  "includes",
  "tags",
]);

export interface ManifestLoadResult {
  manifest: ProfileManifest;
  warnings: ResolutionWarning[];
}

/**
 * Load and validate a profile.json. Returns the parsed manifest plus any
 * non-fatal warnings (R36 unknown-field warnings, missing-manifest warnings).
 *
 * Behavior:
 *  - Missing profile.json → empty manifest + MissingManifest warning. R35
 *    permits all fields to be optional, so a missing manifest is treated as
 *    "default everything" rather than an abort.
 *  - Unparseable JSON → throws InvalidManifestError.
 *  - Wrong types (e.g. extends as number) → throws InvalidManifestError.
 *  - Unknown fields → kept out of the returned manifest, surfaced as warnings.
 */
export async function loadManifest(
  profileDir: string,
  source: string,
): Promise<ManifestLoadResult> {
  const manifestPath = path.join(profileDir, "profile.json");
  const warnings: ResolutionWarning[] = [];

  let raw: string;
  try {
    raw = await fs.readFile(manifestPath, "utf8");
  } catch (err) {
    const e = err as NodeJS.ErrnoException;
    if (e.code === "ENOENT") {
      warnings.push({
        code: "MissingManifest",
        message: `No profile.json at ${manifestPath}; using defaults`,
        source,
      });
      return { manifest: {}, warnings };
    }
    throw new InvalidManifestError(manifestPath, e.message);
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (err) {
    const e = err as Error;
    throw new InvalidManifestError(manifestPath, `JSON parse error: ${e.message}`);
  }

  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    throw new InvalidManifestError(
      manifestPath,
      "expected a JSON object at the root",
    );
  }

  const obj = parsed as Record<string, unknown>;
  const manifest: ProfileManifest = {};

  if (obj["name"] !== undefined) {
    if (typeof obj["name"] !== "string") {
      throw new InvalidManifestError(manifestPath, '"name" must be a string');
    }
    manifest.name = obj["name"];
  }
  if (obj["description"] !== undefined) {
    if (typeof obj["description"] !== "string") {
      throw new InvalidManifestError(manifestPath, '"description" must be a string');
    }
    manifest.description = obj["description"];
  }
  if (obj["extends"] !== undefined) {
    if (typeof obj["extends"] !== "string") {
      throw new InvalidManifestError(manifestPath, '"extends" must be a string');
    }
    manifest.extends = obj["extends"];
  }
  if (obj["includes"] !== undefined) {
    if (
      !Array.isArray(obj["includes"]) ||
      !obj["includes"].every((x): x is string => typeof x === "string")
    ) {
      throw new InvalidManifestError(
        manifestPath,
        '"includes" must be an array of strings',
      );
    }
    manifest.includes = obj["includes"];
  }
  if (obj["tags"] !== undefined) {
    if (
      !Array.isArray(obj["tags"]) ||
      !obj["tags"].every((x): x is string => typeof x === "string")
    ) {
      throw new InvalidManifestError(manifestPath, '"tags" must be an array of strings');
    }
    manifest.tags = obj["tags"];
  }

  for (const key of Object.keys(obj)) {
    if (!KNOWN_FIELDS.has(key)) {
      warnings.push({
        code: "UnknownManifestField",
        message: `Unknown field "${key}" in ${manifestPath}`,
        source,
      });
    }
  }

  return { manifest, warnings };
}

File: go/internal/resolver/manifest.go
package resolver

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"

	pipelineerrors "github.com/htxryan/c3p/internal/errors"
)

var knownManifestFields = map[string]struct{}{
	"name":        {},
	"description": {},
	"extends":     {},
	"includes":    {},
	"tags":        {},
}

// ManifestLoadResult is the loadManifest return: parsed manifest + warnings.
type ManifestLoadResult struct {
	Manifest ProfileManifest
	Warnings []ResolutionWarning
}

// LoadManifest reads and validates profile.json under profileDir. Returns
// the parsed manifest plus any non-fatal warnings (R36 unknown-field
// warnings, missing-manifest warnings).
//
// Behavior matches src/resolver/manifest.ts:
//   - Missing profile.json → empty manifest + MissingManifest warning.
//   - Unparseable JSON → InvalidManifestError.
//   - Wrong types → InvalidManifestError.
//   - Unknown fields → kept out of the returned manifest, surfaced as warnings.
func LoadManifest(profileDir, source string) (ManifestLoadResult, error) {
	manifestPath := filepath.Join(profileDir, "profile.json")
	warnings := []ResolutionWarning{}

	raw, err := os.ReadFile(manifestPath)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			warnings = append(warnings, ResolutionWarning{
				Code:    WarningMissingManifest,
				Message: fmt.Sprintf("No profile.json at %s; using defaults", manifestPath),
				Source:  source,
			})
			return ManifestLoadResult{Manifest: ProfileManifest{}, Warnings: warnings}, nil
		}
		return ManifestLoadResult{}, pipelineerrors.NewInvalidManifestError(manifestPath, err.Error())
	}

	var parsed any
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return ManifestLoadResult{}, pipelineerrors.NewInvalidManifestError(
			manifestPath,
			fmt.Sprintf("JSON parse error: %s", err.Error()),
		)
	}

	obj, ok := parsed.(map[string]any)
	if !ok {
		return ManifestLoadResult{}, pipelineerrors.NewInvalidManifestError(
			manifestPath,
			"expected a JSON object at the root",
		)
	}

	manifest := ProfileManifest{}

	if v, present := obj["name"]; present {
		s, ok := v.(string)
		if !ok {
			return ManifestLoadResult{}, pipelineerrors.NewInvalidManifestError(manifestPath, `"name" must be a string`)
		}
		manifest.Name = s
	}
	if v, present := obj["description"]; present {
		s, ok := v.(string)
		if !ok {
			return ManifestLoadResult{}, pipelineerrors.NewInvalidManifestError(manifestPath, `"description" must be a string`)
		}
		manifest.Description = s
	}
	if v, present := obj["extends"]; present {
		s, ok := v.(string)
		if !ok {
			return ManifestLoadResult{}, pipelineerrors.NewInvalidManifestError(manifestPath, `"extends" must be a string`)
		}
		manifest.Extends = s
	}
	if v, present := obj["includes"]; present {
		arr, ok := v.([]any)
		if !ok {
			return ManifestLoadResult{}, pipelineerrors.NewInvalidManifestError(manifestPath, `"includes" must be an array of strings`)
		}
		out := make([]string, 0, len(arr))
		for _, x := range arr {
			s, ok := x.(string)
			if !ok {
				return ManifestLoadResult{}, pipelineerrors.NewInvalidManifestError(manifestPath, `"includes" must be an array of strings`)
			}
			out = append(out, s)
		}
		manifest.Includes = out
	}
	if v, present := obj["tags"]; present {
		arr, ok := v.([]any)
		if !ok {
			return ManifestLoadResult{}, pipelineerrors.NewInvalidManifestError(manifestPath, `"tags" must be an array of strings`)
		}
		out := make([]string, 0, len(arr))
		for _, x := range arr {
			s, ok := x.(string)
			if !ok {
				return ManifestLoadResult{}, pipelineerrors.NewInvalidManifestError(manifestPath, `"tags" must be an array of strings`)
			}
			out = append(out, s)
		}
		manifest.Tags = out
	}

	for key := range obj {
		if _, known := knownManifestFields[key]; !known {
			warnings = append(warnings, ResolutionWarning{
				Code:    WarningUnknownManifestField,
				Message: fmt.Sprintf("Unknown field %q in %s", key, manifestPath),
				Source:  source,
			})
		}
	}

	return ManifestLoadResult{Manifest: manifest, Warnings: warnings}, nil
}
